File: david/sessions/2020/easy/136_single_number.py
'''
class Solution:
    def singleNumber(self, nums: List[int]) -> int:
        s = set()
        
        for num in nums:
            if num not in s:
                s.add(num)
            else:
                s.remove(num)
        
        return list(s)[0]
'''

from functools import reduce

# O(n)
class Solution:
    def singleNumber(self, nums: List[int]) -> int:
        return reduce(lambda x, y: x ^ y, nums)

File: pia/2020/easy/136_single_number.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var singleNumber = function (nums) {
  let map = {};
  let result = 0;
  for (let i = 0; i < nums.length; i++) {
    const num = nums[i];
    if (!map[num]) {
      map[num] = 1;
    } else {
      map[num]++;
    }
  }
  for (let key in map) {
    if (map[key] === 1) {
      return key;
    }
  }
  return -1;
};
